### PR TITLE
BigQueryInsertJobOperator: log transient error and check job state before marking task as success

### PR DESCRIPTION
**Pull request by pankajastro**
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: providers/src/airflow/providers/google/cloud/operators/bigquery.py [hunk 1]
```diff
@@ -2592,8 +2592,15 @@ def _submit_job(
             nowait=True,
         )
 
-    @staticmethod
-    def _handle_job_error(job: BigQueryJob | UnknownJob) -> None:
+    def _handle_job_error(self, job: BigQueryJob | UnknownJob) -> None:
+        self.log.debug("Job %s is completed. checking the job status", self.job_id)
```
Comment (ephraimbuddy) [inline providers/src/airflow/providers/google/cloud/operators/bigquery.py]:
```suggestion
        self.log.debug("Job %s is completed. Checking the job status", self.job_id)
```

**Review by ephraimbuddy — APPROVED** [review]

**Comment by rawwar** [comment]
There's another active PR: https://github.com/apache/airflow/pull/40863

**Comment by kandharvishnu** [comment]
Closes this issue - https://github.com/apache/airflow/issues/40839

**Comment by pankajastro** [comment]
> There's another active PR: #40863

Yeah, but it has some open discussions, and it's also a pretty old PR. so I thought maybe a new PR would easy.

File: providers/src/airflow/providers/google/cloud/operators/bigquery.py [hunk 2]
```diff
@@ -2592,8 +2592,15 @@ def _submit_job(
             nowait=True,
         )
 
-    @staticmethod
-    def _handle_job_error(job: BigQueryJob | UnknownJob) -> None:
+    def _handle_job_error(self, job: BigQueryJob | UnknownJob) -> None:
+        self.log.info("Job %s is completed. Checking the job status", self.job_id)
+        # I've noticed that sometimes BigQuery jobs transiently report the wrong status, causing the Airflow job to be incorrectly marked as successful.
+        # To avoid this, we refresh the job properties before checking the final state and handling any errors.
+        while job.state != "DONE":
```
Comment (eladkal) [inline providers/src/airflow/providers/google/cloud/operators/bigquery.py]: I don't follow why we should do this.
Is there a bug report to BigQuery that we can track?

I am very against such permanent workarounds. Airflow should not be the fixer for upstream library bugs. Its OK to do this a temporary workaround when upstream library is engaged and working on a fix but otherwise not.

**Review by eladkal — DISMISSED** [review]

File: providers/src/airflow/providers/google/cloud/operators/bigquery.py (hunk 2 above)
Comment (pankajastro) [reply providers/src/airflow/providers/google/cloud/operators/bigquery.py]: Hi @eladkal,

Thank you so much for your review and feedback; I really appreciate it. I understand your concerns regarding the handling of third-party library issues in Airflow, and I agree that addressing it in this manner might not be the ideal approach.

The issue we encountered is quite unusual. We experienced it in production, where the job ultimately timed out after 6 hours, but the `job.result()` method returned before the job had actually finished.

I suspect this could be related to network issues or temporary service disruptions that might have caused the j`ob.result()` to return prematurely. 

Additionally, I feel that checking `job.error_result` may not always be sufficient to detect all errors. As noted in the BigQuery documentation, error_result may not be present in every error response from BigQuery.
https://cloud.google.com/bigquery/docs/error-messages

As for the opening issue in the upstream library, I'm hesitant to pursue it at this stage. I don't have a consistent way to reproduce it, nor enough evidence to definitively say that the problem lies with the BigQuery client or BigQuery service itself.

To address the error handling, would it make sense to modify the `_handle_job_error` method to log any errors in `job.errors` and fail the job if `job.state != "DONE"`? Below is the proposed update:

```python
    def _handle_job_error(self, job: BigQueryJob | UnknownJob) -> None:
        self.log.info("Job %s is completed. Checking the job status", self.job_id)
        # Log any transient errors encountered during the job execution
        for error in job.errors or []:
            self.log.error("BigQuery Job Error: %s", error)
        if job.error_result:
            raise AirflowException(f"BigQuery job {job.job_id} failed: {job.error_result}")
        # Check the final state.
        if job.state != "DONE":
            raise AirflowException(f"Job failed with state: {job.state}")
``` 
I would really appreciate your thoughts on whether this approach makes sense, or if you have any further suggestions.

Comment (shahar1) [reply providers/src/airflow/providers/google/cloud/operators/bigquery.py]: Hey @pankajastro and thanks for putting more details!

Indeed, according to BigQuery docs - the lack of `error_result` could be considered as an expected behavior:
```
The table does not include all possible HTTP errors or other networking errors. Therefore, don't assume that an error object is present in every error response from BigQuery.
```
Therefore, handling it with an additional `if job.state != "DONE"` could be sufficient for now to ensure that the operator does not report errorneous results.

Applying the `while` loop on the other hand, as currently implmented in @kandharvishnu 's [PR](https://github.com/apache/airflow/blame/8fdf29d39dd315f3d7144b4bba66f348221292e6/providers/src/airflow/providers/google/cloud/operators/bigquery.py#L2688), exceeds the atomicity the operation (we already called `job.result()` that should provides a final answer - why do we need to recall it?). That makes us in this case, Airflow maintainers, responsible for handling BQ transient errors - I assume that we don't want to get there.

To sum it up, I'm ok with merging this solution as-is, with Elad's consent.

**Review by shahar1 — APPROVED** [review]

File: providers/src/airflow/providers/google/cloud/operators/bigquery.py (hunk 2 above)
Comment (dejii) [reply providers/src/airflow/providers/google/cloud/operators/bigquery.py]: @shahar1 @pankajastro – I'm suspecting that the `job.state != "DONE"` didn't need to be introduced here. I took a look at the client library while fixing a silent bug introduced by this PR (see fix: https://github.com/apache/airflow/pull/60679), and found that `job.result()` can only return successfully when `state=DONE` or it raises an Exception which will cause the task to fail.

References:
1. https://github.com/googleapis/python-bigquery/blob/73228432a3c821db05d898ea4a4788adf15b033d/google/cloud/bigquery/job/base.py#L990-L1016
2. https://github.com/googleapis/python-bigquery/blob/73228432a3c821db05d898ea4a4788adf15b033d/google/cloud/bigquery/job/query.py#L1770-L1787

Also, per [docs](https://docs.cloud.google.com/bigquery/docs/reference/rest/v2/Job#JobStatus):
> Output only. Running state of the job. Valid states include 'PENDING', 'RUNNING', and 'DONE'.

So it's very unlikely that `"Job failed with state: PENDING|RUNNING"` is a valid exception.

Let me know your thoughts here, I can remove it in my [PR](https://github.com/apache/airflow/pull/60679) since it's somewhat related.
